File: sol/3202/3202.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if n <= 2:
            return n
        for i in range(n):
            nums[i] %= k
        res = 1
        for rem in range(k):
            memo = [0] * k
            for x in nums:
                memo[x] = memo[(rem + k - 1 - x) % k] + 1
            res = max(res, *memo)
        return res

File: sol/3202/3202.cpp
struct Solution {
    int maximumLength(vector<int>& nums, int k) {
        int n = nums.size();
        if (n <= 2){
            return n;
        }
        for (int &x : nums){
            x %= k;
        }
        int res = 1;
        for (int rem = 0; rem < k; ++rem){
            vector<int> memo(k, 0);
            for (int x : nums){
                memo[x] = memo[(rem + k - 1 - x) % k] + 1;
                res = max(res, memo[x]);
            }
        }
        return res;
    }
};
